S2 AvatarGroup

devongovett opened the pull request.

snowystinger commented on packages/@react-spectrum/s2/src/Avatar.tsx:
```diff
@@ -22,20 +22,33 @@ export interface AvatarProps extends StyleProps, DOMProps {
-  src?: string
+  src?: string,
+  /** The stroke color for the avatar. */
+  strokeColor?: 'white' | 'black'
```
is this really staticColor?
Though seems like there's actually a whole bunch of options for the strokeColor. Possibly every "layer" color is a choice?
![Screenshot 2024-08-06 at 10 45 21 AM](https://github.com/user-attachments/assets/2bdb9e2a-b78c-4587-96fe-96e99407ba2e)

snowystinger commented on packages/@react-spectrum/s2/src/Avatar.tsx:
```diff
@@ -60,7 +74,7 @@ function Avatar(props: AvatarProps, ref: DOMRef<HTMLImageElement>) {
       alt={alt}
       style={UNSAFE_style}
-      className={UNSAFE_className + imageStyles(null, props.styles)}
+      className={UNSAFE_className + imageStyles({strokeColor}, props.styles)}
```
```suggestion
      className={(UNSAFE_className ?? '') + imageStyles({strokeColor}, props.styles)}
```

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,93 @@
+  let {children, label, size = 'M'} = props;
+
+  return (
+    <div className={container(null, props.styles)}>
```
need to ass UNSAFE classname and style along with a spread of whatever other props are allowed. (labelling?, id, data-*)

snowystinger replied:
role group?
focusable?

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,93 @@
+
+const container = style({display: 'flex', alignItems: 'center'}, getAllowedOverrides());
+
+export function AvatarGroup(props: AvatarGroupProps) {
```
forward ref?

jluyau commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,163 @@
+    <ul
+      aria-labelledby={labelId}
+      {...filterDOMProps(otherProps)}
+      tabIndex={0}
```
was getting a lint error for this because there's no handler on the list. is that ok?

jluyau replied:
your previous comment mentioned that it should probably be focusable, so i added the tab index for that. was there something else i should've done?

snowystinger replied:
gah, i see how that would be confusing. Sorry about that. Spectrum has been conflicting on this and it's now coming through me. Either way is probably ok for the moment, we'll get it sorted out.

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,163 @@
+      aria-labelledby={labelId}
+      {...filterDOMProps(otherProps)}
+      tabIndex={0}
+      role="group"
```
why are we using an implicit list and converting it to a group? should we just be using divs? or should their be a wrapper element that is the group, and inside is a list and a span label thing?

snowystinger replied:
ah, interesting https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/group_role
This is i think pretty specific to menu implementations

I don't know if it has further implications, would be best to consult with @majornista about this

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,163 @@
+  } = props;
+  let {styles, strokeColor} = useContext(AvatarGroupContext);
+  return (
+    <li className={style({display: 'flex'})}>
```
why display flex?

jluyau replied:
without flex, the list item element looks like this:
![Screenshot 2024-08-09 at 2 10 05 PM](https://github.com/user-attachments/assets/5dccb32e-c20f-4865-92cb-4da0a2619b69)

definitely felt weird using it to fix the styling, so definitely open to another solution

jluyau replied:
i had an issue with the dimensions of the avatars in the `<li>` element, but since i've removed it, it works fine without the flex

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,163 @@
+      XXL: 36,
+      XXXL: 40
+    }
+  }
```
should probably set
`flexShrink: 0` and `flexGrow: 0` on avatars

snowystinger reviewed (commented):
should the '145 members' trailing label truncate?

jluyau commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,163 @@
+          </AvatarGroupContext.Provider>
+        ))
+      }
+      {label && <span id={labelId} className={text({size})}>{label}</span>}
```
i think devon had brought it up in a past Spectrum sync, but i haven't heard if they gave any feedback on how the label should relate to the displayed avatars/whole component

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -77,6 +76,9 @@ const avatar = style({
 const text = style({
   marginStart: 8,
   fontFamily: 'sans',
+  textOverflow: 'ellipsis',
```
you should just be able to use `truncate` to handle this

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -114,32 +116,28 @@ const AvatarGroupContext = createContext<AvatarGroupContextProps>({});
+      <div
+        id={groupId}
+        aria-labelledby={`${groupId} ${labelId}`}
+        aria-label="online members"
```
this should be a prop, we don't know if it'll always be online members

jluyau replied:
ah whoops, used this to test and accidentally committed it. will remove

majornista replied:
If the containing group with `id={groupID}` does not have an `aria-label`, or some default label, like `Avatar Group`, we should not include `groupId` in the `aria-labelledby`, otherwise the group will include the accessibility name for all the descendants, including the label in the accessibility name, which may lead to redundancy and verbosity.

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,156 @@
+  margin: 0
+}, getAllowedOverrides());
+
+function AvatarGroup(props: AvatarGroupProps) {
```
we're not passing through the ref, something like so
```suggestion
function AvatarGroup(props: AvatarGroupProps, ref: ForwardedRef<HTMLElement>) {
```

jluyau replied:
whoops did the first forward but forgot to add/connect the ref 😢

snowystinger commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,156 @@
+import type {StyleString} from '../style/types';
+
+
+interface AvatarGroupProps extends StyleProps, DOMProps {
```
```suggestion
interface AvatarGroupProps extends StyleProps, DOMProps, AriaLabelingProps {
```

devongovett commented:
Pushed 3 changes:

1. Moved to using `Avatar` directly instead of `AvatarGroupItem`. Also moved the stroke into Avatar itself (via an `isOverBackground` prop), since this will also be needed in other places such as Cards.
2. Moved to pixel-based sizing rather than t-shirt sizing (since there are so many size options). Also added back the `size` prop to `Avatar` instead of using macros for sizing, which enables us to adjust the stroke width depending on the size. Updated codemod for this change as well.
3. Made the stroke color automatic. This is done by making all container components with a background color set the `--s2-container-bg` variable. Then the stroke can automatically match the nearest container.

reidbarber reviewed (approved):
LGTM

LFDanLu commented on packages/@react-spectrum/s2/src/AvatarGroup.tsx:
```diff
@@ -0,0 +1,98 @@
+import {useDOMRef} from '@react-spectrum/utils';
+import {useSpectrumContextProps} from './useSpectrumContextProps';
+
+export interface AvatarGroupProps extends UnsafeStyles, DOMProps {
```
Noticed that the aria labeling props were added then removed, any particular reason why?

devongovett replied:
hmm looks like they should, they do get used. I can add them back
